lib/src/dictionary.dart

```diff
@@ -0,0 +1,524 @@
+  final K first;
+  final V second;
+  const _Tuple2(this.first, this.second);
+}
+
+class _EmptyDictionary<K, V> extends Dictionary<K, V> {
+  static const String _rep = '()';
+
+  const _EmptyDictionary() : super._();
+
+  @override
+  Iterable<K> get keys => const [];
+
+  @override
+  Iterable<V> get values => const [];
+
+  @override
+  int get size => 0;
+
+  @override
+  bool containsKey(K _) => false;
+
+  @override
+  String toString() => _rep;
+
+  @override
+  Dictionary<K, V> remove(K _) => this;
+
+  @override
+  V operator [](K _) => null;
+
+  @override
+  Dictionary<K, V> merge(Dictionary<K, V> other) => other;
+
+  @override
+  Dictionary<K, V> assoc(K key, V value) {
+    final root = new BitmapIndexNode<K,V>.empty(0)
+      .assoc(0, key.hashCode, key, value);
+    return new _NodeDictionary._fromRoot(root, 1);
+  }
+
+  @override
+  Iterable<_Tuple2<K, V>> get _pairs => const [];
+
+  @override
+  void forEach(void f(K key, V value)) {}
+}
+
+class _NodeDictionary<K, V> extends Dictionary<K, V> {
+  final Node<K,V> _root;
+
+  @override
+  final int size;
+
+  _NodeDictionary._fromRoot(this._root, this.size): super._();
+
+  @override
+  Iterable<K> get keys => _root.traverseKeys();
+
+  @override
+  Iterable<V> get values => _root.traverseValues();
+
+
+  @override
+  bool containsKey(K key) => _root.find(key.hashCode, key) != null;
+
+  @override
+  Dictionary<K, V> remove(K key) {
+    final newNode = _root.remove(0, key.hashCode, key);
+    if (newNode != _root) {
+      return new _NodeDictionary._fromRoot(newNode, size - 1);
+    }
+    return this;
+  }
+
+  @override
+  V operator [](K key) {
+    return _root.find(key.hashCode, key);
+  }
+
+  @override
+  Dictionary<K, V> merge(Dictionary<K, V> other) => other;
+
+
+  @override
+  Dictionary<K, V> assoc(K key, V value) {
+    final newRoot = _root.assoc(0, key.hashCode, key, value);
+    return new _NodeDictionary._fromRoot(newRoot, size + 1);
+  }
+
+  @override
+  Iterable<_Tuple2<K, V>> get _pairs => _root.traverseBoth();
+
+  @override
+  void forEach(void f(K key, V value)) {
+    for (final tuple in _pairs) {
+      f(tuple.first, tuple.second);
+    }
+  }
+}
+
+/// bit magic
+int mask(int hash, int shift) => ((hash & 0xFFFFFFFF) >> shift) & 0x01F;
+
+/// bit magic
+int bitpos(int hash, int shift) => 1 << mask(hash, shift);
+
+/// bit magic
+int popcount(int i) {
+  i = i - ((i >> 1) & 0x55555555);
+  i = (i & 0x33333333) + ((i >> 2) & 0x33333333);
+  return (((i + (i >> 4)) & 0x0F0F0F0F) * 0x01010101) >> 24;
+}
+
+/// [Node] is an interface for building hash tries.
+abstract class Node<K, V> {
+  /// retrieves the value associated with the given [key], or null.
+  V find(int hash, K key);
+
+  /// associates [key] and [value] with the trie.
+  Node<K, V> assoc(int shift, int hash, K key, V value);
+
+  /// removes [key] from the trie, if present.
+  Node<K, V> remove(int shift, int hash, K key);
+
+  ///
+  Iterable<_Tuple2<K, V>> traverseBoth();
```

Comment:
Since these will be private, I'll keep the names for now but keep it in mind for public interfaces